Adding Workflow Interface FedProx Example with Synthetic Dataset

ParthMandaliya opened the pull request:
DESCRIPTION:
Workflow Interface Tutorial for FedProx with Synthetic non-IID Dataset.

SUMMARY:
Validated the Tutorial for following scenarios:
1. Executed for 30 Total Collaborators, 10 Selected Collaborators, non-IID, and 200 Rounds
![image](https://user-images.githubusercontent.com/25680326/220354609-7aad102d-350b-4e73-9d91-f10918cd3ecc.png)

2. Executed for 100 Total Collaborators, 20 Selected Collaborators, non-IID, and 200 Rounds
![image](https://user-images.githubusercontent.com/25680326/220355076-07ffeedf-0457-4ace-942f-84e036f953cc.png)

3. Executed for 30 Total Collaborators, 10 Selected Collaborators, IID, and 200 Rounds
![image](https://user-images.githubusercontent.com/25680326/220355372-6e6da382-54a2-422d-a081-d18d09679a17.png)

4. Validated in Google Collab

grib0ed0v committed:
Update README.md (#719)

Change github Python badge's

Signed-off-by: Alexey Gruzdev <a.s.gruzdev@yandex.ru>
Signed-off-by: ParthM-GitHub <parthx.mandaliya@intel.com>

operepel committed:
Add files via upload

Picture for the Contributing.md

Signed-off-by: ParthM-GitHub <parthx.mandaliya@intel.com>

operepel committed:
Add files via upload

Signed-off-by: ParthM-GitHub <parthx.mandaliya@intel.com>

operepel committed:
Update CONTRIBUTING.md

Signed-off-by: ParthM-GitHub <parthx.mandaliya@intel.com>

operepel committed:
Update CONTRIBUTING.md

Signed-off-by: ParthM-GitHub <parthx.mandaliya@intel.com>

operepel committed:
Update README.md

Signed-off-by: ParthM-GitHub <parthx.mandaliya@intel.com>

psfoley committed:
Fixed workflow interface notebook requirements (#729)

Signed-off-by: Patrick Foley <psfoley@gmail.com>
Signed-off-by: ParthM-GitHub <parthx.mandaliya@intel.com>

ParthM-GitHub committed:
Adding Workflow Interface FedProx Example with Synthetic Dataset

Signed-off-by: ParthM-GitHub <parthx.mandaliya@intel.com>

operepel committed:
Update CONTRIBUTING.md

Signed-off-by: ParthM-GitHub <parthx.mandaliya@intel.com>

operepel committed:
Update CONTRIBUTING.md

Signed-off-by: ParthM-GitHub <parthx.mandaliya@intel.com>

itrushkin committed:
Fix CONTINUE_GLOBAL optimizer treatment (#711)

* Fix variable name

Signed-off-by: Ilya Trushkin <ilya.trushkin@intel.com>

* Use CONTINUE_GLOBAL treatment in FedProx tutorial

Signed-off-by: Ilya Trushkin <ilya.trushkin@intel.com>

---------

Signed-off-by: Ilya Trushkin <ilya.trushkin@intel.com>
Signed-off-by: ParthM-GitHub <parthx.mandaliya@intel.com>

ParthM-GitHub committed:
Modified FedProx Example, tested following scenarios
1. 10 Selected Collaborators, 30 Total Collaborators, with 200 Rounds (FedProx vs FedAvg)
2. 20 Selected Collaborators, 100 Total Collaborators, with 200 Rounds (FedProx vs FedAvg)
3. 10 Selected Collaborators, 30 Total Collaborators, with 200 Rounds (FedProx vs FedAvg) Tested on Google Colab

Signed-off-by: ParthM-GitHub <parthx.mandaliya@intel.com>

ParthM-GitHub committed:
Uncommented few necessary lines

Signed-off-by: ParthM-GitHub <parthx.mandaliya@intel.com>

ParthM-GitHub committed:
Updated FedProx example

Signed-off-by: ParthM-GitHub <parthx.mandaliya@intel.com>

ParthM-GitHub committed:
Updated FedProx example

Signed-off-by: ParthM-GitHub <parthx.mandaliya@intel.com>

ParthM-GitHub committed:
Updated FedProx example

Signed-off-by: ParthM-GitHub <parthx.mandaliya@intel.com>

ParthM-GitHub committed:
Update Workflow_Interface_401_FedProx_with_Synthetic_nonIID.ipynb

Signed-off-by: ParthM-GitHub <parthx.mandaliya@intel.com>

ParthM-GitHub committed:
Changes as per numpy 1.24.x

Made changes required to run the example with numpy 1.24.x version.

Signed-off-by: ParthM-GitHub <parthx.mandaliya@intel.com>

mansishr commented on openfl-tutorials/experimental/Workflow_Interface_401_FedProx_with_Synthetic_nonIID.ipynb:
```diff
@@ -0,0 +1,802 @@
+    "        This is the 'end' step.\n",
+    "        \"\"\"\n",
+    "        self.round_number += 1\n",
+    "        print('This is end of the flow')"
```
This print statement is misleading because it gets printed after every round. Let's pass `total_rounds` to federated flow and print this condition once at the end of the flow.

ParthM-GitHub replied:
Incorporated

mansishr commented on openfl-tutorials/experimental/Workflow_Interface_401_FedProx_with_Synthetic_nonIID.ipynb:
```diff
@@ -0,0 +1,802 @@
+    "    return accuracy, test_loss, correct\n",
+    "\n",
+    "\n",
+    "def weighted_average(tensors, weights):\n",
```
I propose that we create `openfl.experimental.interface.aggregation_functions` and move this function there.

ParthM-GitHub replied:
Incorporated

mansishr commented on openfl-tutorials/experimental/Workflow_Interface_401_FedProx_with_Synthetic_nonIID.ipynb:
```diff
@@ -0,0 +1,802 @@
+   "source": [
+    "!pip install git+https://github.com/intel/openfl.git\n",
+    "!pip install -r https://raw.githubusercontent.com/intel/openfl/develop/openfl-tutorials/experimental/requirements_workflow_interface.txt\n",
+    "!pip install matplotlib\n",
```
!pip install seaborn is missing

ParthM-GitHub replied:
Incorporated

ParthM-GitHub committed:
Added following changes:

1. Converted flow to internal loop flow
2. Added seaborn library install command

Signed-off-by: ParthM-GitHub <parthx.mandaliya@intel.com>

ParthM-GitHub committed:
Merge branch 'intel:develop' into fedprox_example

ParthM-GitHub committed:
Refactored WeightedAverage:
1. PyTorch: openfl/experimental/interface/keras/aggregation_functions/weighted_average.py
2. Keras: openfl/experimental/interface/keras/aggregation_functions/weighted_average.py (with NotImplementedError)
3. PyTorch Weighted average testcase in tests/openfl/experimental/test_weighted_average.py

Signed-off-by: ParthM-GitHub <parthx.mandaliya@intel.com>

ParthM-GitHub committed:
Added __init__.py file in tests/openfl/experimental folder.

Signed-off-by: ParthM-GitHub <parthx.mandaliya@intel.com>

ParthM-GitHub committed:
Fixing lint suggestions

Signed-off-by: ParthM-GitHub <parthx.mandaliya@intel.com>

ParthM-GitHub committed:
Fixing lint suggestions.

Signed-off-by: ParthM-GitHub <parthx.mandaliya@intel.com>

ParthM-GitHub committed:
Fixing lint suggestions.

Signed-off-by: Parth Mandaliya <parthx.mandaliya@intel.com>

ParthM-GitHub committed:
Fixing lint suggestions.

Signed-off-by: Parth Mandaliya <parthx.mandaliya@intel.com>

ParthM-GitHub committed:
Adding optimizer.pth file for optimizer weighted average test case.

Signed-off-by: Parth Mandaliya <parthx.mandaliya@intel.com>

ParthM-GitHub committed:
Merge branch 'securefederatedai:develop' into fedprox_example

ParthM-GitHub committed:
Merge branch 'securefederatedai:develop' into fedprox_example

ParthM-GitHub committed:
Merge branch 'securefederatedai:develop' into fedprox_example

ParthM-GitHub committed:
Merge branch 'securefederatedai:develop' into fedprox_example

theakshaypant commented:
@ParthMandaliya can you please resolve the comments/conflicts and request for review again?

ParthMandaliya commented:
@theakshaypant This 401 FedProx tutorial has already been merged via a different PR, and is available under openfl-tutorials/experimental/workflow directory.

theakshaypant commented:
> @theakshaypant This 401 FedProx tutorial has already been merged via a different PR, and is available under openfl-tutorials/experimental/workflow directory.
> 
> cc: @ishant162

Ok, thank you! Closing this PR in that case.
